### DISCUSSION TITLE
Package lookup failures

### DISCUSSION THREAD

**Discussion by dawidhyzy** [renovatebot/renovate#24392] [Request Help]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

36.82.0

### If you're self-hosting Renovate, select which platform you are using.

Bitbucket Cloud (bitbucket.org)

### Was this something which used to work for you, and then stopped?

I am trying to get this working for the first time

### Describe the problem

After adding a new dependency to our project that is hosted on `https://artifact.plugin.npaw.com/artifactory/plugins/android` we started getting a lookup failure warning. 

I see the warning in the logs saying:

```
 WARN: Package lookup failures (repository=repo/project)
              "warnings": [
                    "Failed to look up maven package com.npaw.plugin:plugin-media3-exoplayer",
                     "Failed to look up maven package com.npaw.plugin:plugin"
              ],
             "files": ["gradle/libs.versions.toml"]
```       
I'm not sure if that's the artifact repository problem or it's not picked up from my `settings.gradle.kts` file: 

```
dependencyResolutionManagement {
    repositoriesMode.set(RepositoriesMode.FAIL_ON_PROJECT_REPOS)
    repositories {
        google()
        mavenCentral()
        maven("https://artifact.plugin.npaw.com/artifactory/plugins/android")
    }
}
```

The outcome is that I don't get any updates for these dependencies even though I can see that there were a couple of new versions.

Is there anything I could do on my end to fix that?

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste the relevant log(s) here, between the starting and ending backticks
```

</details>

### Have you created a minimal reproduction repository?

I have explained in the description why a minimal reproduction is impossible

**Comment by rarkins** [comment]
First you need to consult the debug logs to be sure that renovate is attempting the right registry/URLs

**Comment by dawidhyzy** [comment]
I see that it looks for other dependencies in this registry

```
DEBUG: Looking up org.junit.jupiter:junit-jupiter in repository https://artifact.plugin.npaw.com/artifactory/plugins/android/
DEBUG: GET https://artifact.plugin.npaw.com/artifactory/plugins/android/com/google/truth/truth/maven-metadata.xml = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=284)
```

**Comment by dawidhyzy** [comment]
It looks for the other desired dependency too
```
DEBUG: Looking up com.npaw.plugin:plugin-media3-exoplayer in repository https://artifact.plugin.npaw.com/artifactory/plugins/android/
DEBUG: Found 12 new releases for com.chimerapps.niddler:niddler in repository https://repo.maven.apache.org/maven2/
DEBUG: Looking up com.chimerapps.niddler:niddler in repository https://artifact.plugin.npaw.com/artifactory/plugins/android/
DEBUG: GET https://artifact.plugin.npaw.com/artifactory/plugins/android/org/mockito/mockito-inline/maven-metadata.xml = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=278)
DEBUG: Content is not found for Maven url
       "url": "https://artifact.plugin.npaw.com/artifactory/plugins/android/org/mockito/mockito-inline/maven-metadata.xml",
       "statusCode": undefined
```

**Reply by rarkins** [reply]
I don't understand why you're mentioning other dependencies instead of the one you originally asked about.

The point of my question - which is not fully answered yet - is to know: is Renovate looking in the right place? If so then it becomes an auth problem, not a URL problem.

**Reply by dawidhyzy** [reply]
The reason why I'm mentioning it is that taking into account the order of the repositories declaration they should be already found in `google` or `mavenCentral` and `https://artifact.plugin.npaw.com/artifactory/plugins/android` could be skipped.

```
repositories {
        google()
        mavenCentral()
        maven("https://artifact.plugin.npaw.com/artifactory/plugins/android")
}
```
As far as I understand the logs it does:
`DEBUG: Looking up com.npaw.plugin:plugin-media3-exoplayer in repository https://artifact.plugin.npaw.com/artifactory/plugins/android/`
Am I right? Sorry, I don't have a knowledge how the process looks like.

**Reply by rarkins** [reply]
Thanks for explaining, glad I asked. Yes, this is because the Maven datasource today uses "merge" strategy, which means it tries all registries: https://github.com/renovatebot/renovate/blob/c5aae245caad3288fd4e44519c70e8f5c87e7b12/lib/modules/datasource/maven/index.ts#L69

Associated bug: https://github.com/renovatebot/renovate/issues/22744

I think we'll fix it in the next major release, because it _will_ change some results even though they should in theory all be for the better.

So this confirms that Renovate is aware of and trying your custom URL, which implies that the problem is _authentication_, which can be fixed with `hostRules`. Usually this is either Bearer auth (a simple `token` or Basic auth (either token with authType=basic, or username/password). Which type of auth does Artifactory give you to use?

**Reply by dawidhyzy** [reply]
Thanks for explaining that!

Frankly, it allows fetching dependencies without authentication(I didn't add any credentials to do that). And I can[ browse it](https://artifact.plugin.npaw.com/artifactory/plugins/android/com/npaw/plugin/plugin-media3-exoplayer/) manually as well. I could ask the owners about it but I don't know what Renovate is trying to call that requires authentication. Any pointers on how to find that?

**Reply by rarkins** [reply]
In that case it's maybe _not_ authentication, but instead that this URL is not found:

```
DEBUG: GET https://artifact.plugin.npaw.com/artifactory/plugins/android/com/google/truth/truth/maven-metadata.xml = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=284)
```

So the question is why the maven-metadata.xml file is not found

**Reply by dawidhyzy** [reply]
I think the reason is the `merge` strategy because this registry doesn't contain `com.google.truth:truth` and we want to look it up only for `com.npaw.plugin:plugin` & `com.npaw.plugin:plugin-media3-exoplayer`
